### DISCUSSION TITLE
Enhanced JavaScript module from this repository

### DISCUSSION THREAD

**Discussion by jerinphilip** [browsermt/bergamot-translator#433] [Ideas]
The current JavaScript "[example](https://github.com/browsermt/bergamot-translator/blob/b1e5a48f1a3e3a6d8a6d553dafe32d881eb3aabe/wasm/test_page/js/worker.js)"  for the test page is perhaps not the best provision from here. There are a few common functions that can be abstracted here, potentially tested headless and seen by multiple eyes. Therefore, a better solution is a JavaScript library or module, ideally with an API consistent with the C++ from which it is derived. 

The following are ideas of a few downstream applications of a JavaScript library that can do translation and is batteries included: 

1. The extension could be a consumer of this tested library eliminating human errors that come out of manual copying. (We got `alignment:true` and `responseOptions.HTML` correct here, but transfer to [firefox-translations](https://github.com/mozilla/firefox-translations/) led to inconsistent config).
2. https://github.com/jerinphilip/lemonade/issues/56.  
3. A plugin that can be embedded on to sites to provide translation similar to a google offering: https://www.dummies.com/article/technology/notable-websites/google/how-to-install-google-translates-website-translator-plugin-145074

There are parts of this library already existing in sources across a few repositories. At the moment the extra work and duplication is convenient for faster development cycles. But at some point in the future, we should look at consolidating commonly used functionalities here co-located with the bindings and continuously tested. The JavaScript module can be part of the releases from this repository. 

This thread is meant to investigate and discuss ideas, weigh the pros and cons of the approach.

**Comment by jerinphilip** [comment]
OpenCV is a mature C++ library which exposes bindings to JavaScript and has JavaScript tests (https://docs.opencv.org/3.4/d4/da1/tutorial_js_setup.html) which we could potentially draw references/examples from. They also appear to be using WebAssembly SIMD and threads.

They allow automated testing variations of:

i. Manually in browser
ii. Headless with Puppeteer
iii. Using Node.js.

**Comment by jerinphilip** [comment]
There appears to be a headless mechanism to test JavaScript via Firefox of which an example can be found in https://github.com/mozilla/firefox-translations/pull/106. 

An incremental approach could be doing (1) manually in the browser and deploying to some health status Github page. This will prepare the test primitives and then we can look at the tests running headless.

**Comment by jerinphilip** [comment]
@jelmervdl has successfully managed to run the WASM module via node.js in https://gist.github.com/jelmervdl/a4c8b6b92ad88a885e1cbd51c6ad4902. This opens up possibilities for headless testing.

**Comment by jelmervdl** [comment]
I still like to make (3) the embedded version a la Google Translate SDK someday. It would also include (yet another copy of) a model index and a GUI… We could skip over the language detection bit and ask whoever embeds the script in their website to at least tell what language the website needs to be translated from. Presumably they know.

And I don't know what to do about Safari. [No support for SIMD](https://bugs.webkit.org/show_bug.cgi?id=222382)… but also it does already have support for translating websites embedded. But so does Chrome. And Firefox, with the extension. So who exactly is (3) for?

**Reply by jerinphilip** [reply]
First time I saw 3 was some conference using it for their web page. Perhaps this is an accessibility thing (with not much privacy implications? Does google use this JavaScript to do tracking stuff?). I'm just thinking of possible applications. This was an ACCV conference which gave a drop-down button for translation (https://www.accv2018.net/) - looks like they're using what you call a "Google Translate SDK".

> So who exactly is (3) for?

If there's a Google Translate SDK offering, there must be some users. An open privacy minded alternative should have some use in the segment. More usage should drive more bug reports and a fraction which will hopefully convert into maintainence and fixes (I'm being overly optimistic :D). We already have issues like #418, which demands slower but useful features. We can start with what works and expand to Safari eventually? Does Safari use a translation server somewhere or local?

**Reply by jelmervdl** [reply]
Seems like Safari is always sending it to Apple at this moment:
<img width="780" alt="image" src="https://user-images.githubusercontent.com/198639/176561754-da64de64-91bc-44f1-8166-ea0f94644f25.png">

The Privacy & Translation button opens a popup saying: _The text you choose to have translated is sent to Apple to process your requests. Your requests are associated with a random, rotating identifier, not your Apple ID, email address, or other data Apple may have from your use of other Apple services. By default, Apple stores transcripts of your text inputs for two years and may review a subset of these transcripts. The small subset of requests that have been reviewed may be kept beyond two years for ongoing improvement of Translate, Systemwide translation, and other natural language processing features such as Siri._

_Translate should not be relied on in circumstances where you could be harmed or injured, in high-risk situations, for navigation, or for the diagnosis or treatment of any medical condition._

So you're right, I can imagine some websites might want to offer local translations through an embedded javascript.
